ScreenCaptureKitCaptureSource::streamDidOutputVideoSampleBuffer should crop video frames as needed

youennf opened the pull request:
#### e46ffa98cd5d3466abcf9621b01e1b712921ff48
<pre>
ScreenCaptureKitCaptureSource::streamDidOutputVideoSampleBuffer should crop video frames as needed
<a href="https://rdar.apple.com/127598482">rdar://127598482</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=273764">https://bugs.webkit.org/show_bug.cgi?id=273764</a>

Reviewed by Jean-Yves Avenard.

ScreenCaptureKitCaptureSource::streamDidOutputVideoSampleBuffer receives buffers with black stripes.
This happens in case of resizing or when large presenter overlay is enabled.
To fix this, we crop via ImageTransferSessionVT the pixel buffer to remove those black stripes.

We update ImageTransferSessionVT to allow setting a cropping rectangle.
We set kCVImageBufferCleanApertureKey on the source buffer before doing the transfer to the new buffer.

Manually tested by:
- Capture screen, capture camera and go to large presenter overlay
- Capture screen and resize captured window (w/o presenter overlay).

* Source/WebCore/platform/cocoa/CoreVideoSoftLink.cpp:
* Source/WebCore/platform/cocoa/CoreVideoSoftLink.h:
* Source/WebCore/platform/cocoa/VideoToolboxSoftLink.cpp:
* Source/WebCore/platform/cocoa/VideoToolboxSoftLink.h:
* Source/WebCore/platform/graphics/cv/ImageTransferSessionVT.h:
* Source/WebCore/platform/graphics/cv/ImageTransferSessionVT.mm:
(WebCore::ImageTransferSessionVT::setCroppingRectangle):
(WebCore::ImageTransferSessionVT::convertPixelBuffer):
* Source/WebCore/platform/mediastream/mac/ScreenCaptureKitCaptureSource.h:
* Source/WebCore/platform/mediastream/mac/ScreenCaptureKitCaptureSource.mm:
(WebCore::ScreenCaptureKitCaptureSource::streamDidOutputVideoSampleBuffer):

Canonical link: <a href="https://commits.webkit.org/278475@main">https://commits.webkit.org/278475@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/851662e3528398bda17d9385cf0c098801556588

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/50638 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/29935 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/2954 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/53897 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/1329 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/52941 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/36195 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/979 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/41279 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/52737 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/27591 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/43613 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/22393 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/24980 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/870 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/9079 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/46959 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/932 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/55486 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/25740 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/840 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/48692 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/26997 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/43758 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/47747 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11104 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/27865 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/26729 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/platform/graphics/cv/ImageTransferSessionVT.mm:
```diff
@@ -73,6 +73,26 @@
+        return;
+    }
+
+    m_croppingRectangle = rectangle;
```
WTFMove(rectangle)

Can be move above considering this is always done (so no need for `m_croppingRectangle = { };`

youennf replied:
Will do

youennf committed:
ScreenCaptureKitCaptureSource::streamDidOutputVideoSampleBuffer should crop video frames as needed
rdar://127598482
https://bugs.webkit.org/show_bug.cgi?id=273764

Reviewed by Jean-Yves Avenard.

ScreenCaptureKitCaptureSource::streamDidOutputVideoSampleBuffer receives buffers with black stripes.
This happens in case of resizing or when large presenter overlay is enabled.
To fix this, we crop via ImageTransferSessionVT the pixel buffer to remove those black stripes.

We update ImageTransferSessionVT to allow setting a cropping rectangle.
We set kCVImageBufferCleanApertureKey on the source buffer before doing the transfer to the new buffer.

Manually tested by:
- Capture screen, capture camera and go to large presenter overlay
- Capture screen and resize captured window (w/o presenter overlay).

* Source/WebCore/platform/cocoa/CoreVideoSoftLink.cpp:
* Source/WebCore/platform/cocoa/CoreVideoSoftLink.h:
* Source/WebCore/platform/cocoa/VideoToolboxSoftLink.cpp:
* Source/WebCore/platform/cocoa/VideoToolboxSoftLink.h:
* Source/WebCore/platform/graphics/cv/ImageTransferSessionVT.h:
* Source/WebCore/platform/graphics/cv/ImageTransferSessionVT.mm:
(WebCore::ImageTransferSessionVT::setCroppingRectangle):
(WebCore::ImageTransferSessionVT::convertPixelBuffer):
* Source/WebCore/platform/mediastream/mac/ScreenCaptureKitCaptureSource.h:
* Source/WebCore/platform/mediastream/mac/ScreenCaptureKitCaptureSource.mm:
(WebCore::ScreenCaptureKitCaptureSource::streamDidOutputVideoSampleBuffer):

Canonical link: https://commits.webkit.org/278475@main
